### PR TITLE
Large places

### PR DESCRIPTION
**Pull request by vext01**
The first commit message explains.

There's some renaming mixed in with the first commit. I tried to separate it out, but it's more trouble than it's worth. Poor planning on my part. Sorry about that.

### CODE REVIEW

File: ykcompile/src/lib.rs [hunk 1]
```diff
@@ -550,11 +550,12 @@ impl<TT> TraceCompiler<TT> {
                     }
                     _ => todo!(),
                 }
-                return;
+                return Location::Register(reg, false);
```
Comment (jacob-hughes) [inline ykcompile/src/lib.rs:554]: Is the `return` keyword here necessary?

File: ykcompile/src/lib.rs [hunk 2]
```diff
@@ -850,73 +871,179 @@ impl<TT> TraceCompiler<TT> {
...
+                            dynasm!(self.asm
+                                ; mov Rq(temp), QWORD [Rq(src_ro.reg) + src_ro.offs]
+                                ; mov QWORD [Rq(dest_ro.reg) + dest_ro.offs], Rq(temp)
+                            );
+                        }
+                        _ => unreachable!(),
+                    }
+                    self.free_if_temp(Location::Register(temp, false));
+                } else {
+                    self.c_memcpy(dest_ro, src_ro, ty.size());
```
Comment (jacob-hughes) [inline ykcompile/src/lib.rs:972]: Can these locations overlap?

File: ykcompile/src/lib.rs [hunk 3]
```diff
@@ -480,50 +494,35 @@ impl<TT> TraceCompiler<TT> {
...
+        let sym = Self::find_symbol("memcpy").unwrap();
+        self.caller_save();
+        dynasm!(self.asm
+            ; push rax
+            ; xor rax, rax
+            ; lea rdi, [Rq(dest.reg) + dest.offs]
+            ; lea rsi, [Rq(src.reg) + src.offs]
+            ; mov rdx, size as i32
+            ; mov r11, QWORD sym as i64
+            ; call r11
```
Comment (jacob-hughes) [inline ykcompile/src/lib.rs:509]: Why is it necessary to lookup the `memcpy` fn pointer, put it in a register and then call it. Can't you just call memcpy directly here?

Comment (vext01) [reply ykcompile/src/lib.rs:509]: I've been unable to achieve that using `dynasm` so far. Looking at the docs, it seems you can't.

image


Notice there is no `imm64` variant. [This SO post](
https://stackoverflow.com/questions/19552158/call-an-absolute-pointer-in-x86-machine-code) confirms it too.

I think you can use `rel32` [if you know where memcpy is relative to the instruction pointer](https://stackoverflow.com/questions/38961192/how-to-execute-a-call-instruction-with-a-64-bit-absolute-address)

File: ykcompile/src/lib.rs (hunk 1 above)
Comment (vext01) [reply ykcompile/src/lib.rs:554]: Yes, it's an intentional early return. During development I removed it accidentally and hilarity ensued.

File: ykcompile/src/lib.rs (hunk 2 above)
Comment (vext01) [reply ykcompile/src/lib.rs:972]: And thus should we use `memmove()`. Good question.

I don't think they can overlap. @ptersilie?

Comment (vext01) [reply ykcompile/src/lib.rs:972]: Can you guys think of a rust assignment that would copy into memory which overlaps with the source operand? Does the borrow checker disallow that perhaps.

Comment (bjorn3) [reply ykcompile/src/lib.rs:972]: The MIR semantics regarding aliasing are not clear right now.

Comment (ptersilie) [reply ykcompile/src/lib.rs:972]: I honestly don't know. But gut feeling says no, but who knows?

File: ykcompile/src/lib.rs (hunk 3 above)
Comment (jacob-hughes) [reply ykcompile/src/lib.rs:509]: Huh, strange. That stack overflow post was interesting.

File: ykcompile/src/lib.rs (hunk 1 above)
Comment (jacob-hughes) [reply ykcompile/src/lib.rs:554]: Ah yes, I didn't see that it was part of the if block.

File: ykcompile/src/lib.rs (hunk 2 above)
Comment (vext01) [reply ykcompile/src/lib.rs:972]: It sounds like if the buffers overlap, we invoke undefined behaviour. So to be completely safe, I suppose we should use `memmove` and take the performance penalty.

Comment (jacob-hughes) [reply ykcompile/src/lib.rs:972]: Yeah I agree that's the sensible approach.

**Comment by jacob-hughes** [comment]
LGTM. Does this need squashing?

**Comment by vext01** [comment]
If that last push is OK, I'll squash.

**Comment by jacob-hughes** [comment]
Yeah, please squash

**Comment by vext01** [comment]
splat.

**Pull request merged**